frontend/src/index.tsx

```diff
@@ -6,8 +6,50 @@ import "bootstrap/dist/css/bootstrap.min.css";
 import "./index.css";
 import App from "./App";
 import { Provider } from "react-redux";
-import { configureStore, getDefaultMiddleware } from "@reduxjs/toolkit";
+import {
+  configureStore,
+  getDefaultMiddleware,
+  Middleware,
+  AnyAction,
+  MiddlewareAPI,
+} from "@reduxjs/toolkit";
 import { rootReducer } from "./store/reducers";
+import { Dispatch } from "./store";
+import { initialState, RootSchema } from "./store/root-schema";
+import { EventsObserver } from "./store/stats/events-observer";
+
+const eventsObserver = new EventsObserver();
+
+const getStoredStats = () => {
+  return Array.from(Array(localStorage.length).keys())
+    .map(k => localStorage.key(k) as string)
+    .reduce(
+      (prev, cur) => ({
+        ...prev,
+        [cur]: JSON.parse(localStorage.getItem(cur) ?? "{}"),
+      }),
+      {},
+    );
+};
+
+const storedStats = getStoredStats();
+
+const saveStorageMiddleware: Middleware = (store: MiddlewareAPI) => (
+  next: Dispatch,
+) => (action: AnyAction) => {
+  next(action);
+  if (action.type.startsWith("stats/")) {
+    const state = store.getState() as RootSchema;
+    localStorage.setItem(state.username, JSON.stringify(state.stats[state.username]));
```

Comment:
Mmm. I get the idea of saving things under user's username... Potentially useful if there is more than 1 player playing from the same machine, but very confusing, if you think you can choose your nickname, and then e.g. refresh the browser and see your statistics (like I thought you could), and you see nothing, because there is no nickname set.
I'd save statistics under some fixed key "stats", or sth.

Reply:
That's a product decision. If you prefer to have one stats per computer that's fine, it will allow to change nickname but won't allow to share computer. Just write me final decision ;)

Reply:
Both are pretty quick to accomplish

Reply:
I'll reach out to the client quickly.

Reply:
Ok, seems like client prefers "per nickname" as well. :)